### ISSUE TITLE
Unable to update packages using - winget upgrade --all

### ISSUE THREAD

**Issue by smoonlee** [microsoft/winget-cli#4909]
When trying to run `winget upgrade --all` I get the following response 

```
An unexpected error occurred while executing the command:
Bad optional access
```

I've also tried using the latest preview release of winget-cli and get the same error.

```
# winget --version
v1.8.1911
```

```
# winget --version
v1.9.25170
```

> Windows Information
```
OS Name:                       Microsoft Windows 11 Pro
OS Version:                    10.0.26100 N/A Build 26100
OS Manufacturer:               Microsoft Corporation
OS Configuration:              Standalone Workstation
```

**Comment by Trenly** [comment]
Can you share the log file from the Logs folder when you run `winget upgrade --all --verbose --logs` ?

**Comment by smoonlee** [comment]
Well, yes and no. - Not sure if this started after the 24H2 update

 ![Image](https://github.com/user-attachments/assets/243723c3-5fa4-45af-9558-5bf003fac306)

**Comment by Trenly** [comment]
@smoonlee - did it not open the logs folder after the command exited? If it didn't, the logs folder location can be opened with just `winget --logs`, or seen with `winget --info`.

In order to further diagnose the issue, we will likely need the actual log file from the verbose upgrade command

**Comment by smoonlee** [comment]
[WinGet-2024-10-26-10-33-58.549.log](https://github.com/user-attachments/files/17530761/WinGet-2024-10-26-10-33-58.549.log)

Attached :)

**Comment by Trenly** [comment]
Thanks! It looks like something has happened where your pinning database is un-openable; 

The database should be at `%LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\pinning.db`.

My recommendation would be to try renaming that file to something like `pinning-old` or deleting it entirely and seeing if that fixes the issue. This will remove any pins you had added with `winget pin`, though

**Comment by smoonlee** [comment]
Renamed, But still same error.

![Image](https://github.com/user-attachments/assets/5ca1d59a-e508-4880-a728-aca4ccaddac2)

**Comment by avi12** [comment]
I can't even upgrade a single package
```
PowerShell 7.4.5

   A new PowerShell stable release is available: v7.4.6 
   Upgrade now, or check out the release page at:       
     https://aka.ms/PowerShell-Release?tag=v7.4.6       

Loading personal and system profiles took 1472ms.
PS C:\repositories\extensions\youtube-like-dislike-shortcut> winget upgrade Microsoft.PowerShell -h                                                                                                                                  
No installed package found matching input criteria.
```

**Comment by Trenly** [comment]
Thanks for trying!

**Comment by smoonlee** [comment]
> I can't even upgrade a single package
> 
> ```
> PowerShell 7.4.5
> 
>    A new PowerShell stable release is available: v7.4.6 
>    Upgrade now, or check out the release page at:       
>      https://aka.ms/PowerShell-Release?tag=v7.4.6       
> 
> Loading personal and system profiles took 1472ms.
> PS C:\repositories\extensions\youtube-like-dislike-shortcut> winget upgrade Microsoft.PowerShell -h                                                                                                                                  
> No installed package found matching input criteria.
> ```

How did you install PowerShell though? From the Store or winget?

**Comment by avi12** [comment]
@smoonlee Winget

**Comment by smoonlee** [comment]
> [@smoonlee](https://github.com/smoonlee) Winget

What do you get if you run 

```
winget list Microsoft.PowerShell
```

**Comment by avi12** [comment]
```
PS C:\Users\avi12> winget list Microsoft.PowerShell
No installed package found matching input criteria.
```

**Comment by smoonlee** [comment]
What version of winget you got installed?

**Comment by avi12** [comment]
v1.8.1911

**Comment by smoonlee** [comment]
Ok, So I fixed my issue, buy installing all the pending updates manually and it looks to now be working fine again.

**Comment by avi12** [comment]
I mean, I did try to upgrade PowerShell, unsuccessfully
```
PS C:\Users\avi12> winget upgrade Microsoft.PowerShell -h
No installed package found matching input criteria.
```

**Comment by smoonlee** [comment]
What happens if you try 

``` pwsh
winget install --id Microsoft.PowerShell
```

**Comment by avi12** [comment]
```
PS C:\Users\avi12> winget install --id Microsoft.PowerShell -h
Found PowerShell [Microsoft.PowerShell] Version 7.4.6.0
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://github.com/PowerShell/PowerShell/releases/download/v7.4.6/PowerShell-7.4.6-win-x64.msi
  ██████████████████▍             64.1 MB /  104 MB
```

**Comment by smoonlee** [comment]
What happens if you try it without the `-h` flag

**Comment by avi12** [comment]
Never mind, it took a while but eventually the output became
```
PS C:\Users\avi12> winget install --id Microsoft.PowerShell -h
Found PowerShell [Microsoft.PowerShell] Version 7.4.6.0
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://github.com/PowerShell/PowerShell/releases/download/v7.4.6/PowerShell-7.4.6-win-x64.msi
  ██████████████████████████████   104 MB /  104 MB
Successfully verified installer hash
Starting package install...
Successfully installed
```
Interestingly, when I tried to upgrade, this time I got
```
PS C:\Users\avi12> winget upgrade Microsoft.PowerShell -h
No available upgrade found.
No newer package versions are available from the configured sources.
```
but I don't trust the problem to have gone away

**Comment by smoonlee** [comment]
What is you try a source reset?

You'll need to run as Administrator

```
winget source reset --force ; winget source update
```

**Comment by avi12** [comment]
```
PS C:\Users\avi12> winget source reset --force ; winget source update
Resetting all sources...Done
Updating all sources...
Updating source: msstore...
Done
Updating source: winget...
Done
```
not sure if it helped in the long term

**Comment by smoonlee** [comment]
time will tell. :)

**Comment by matthartsock** [comment]
For me, I went to `%localappdata%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe`, opened the folder properties and unchecked the read-only checkbox, then applied to all sub folders and items.  Doing so seems to have let me use the -all command successfully.  However, between this step and trying the --all command, I did run several individual upgrades using the --id flag and specifying objects until I only had about 20 left to update.
